internal/engine/wazevo/frontend/lower.go

```diff
@@ -3829,13 +3829,25 @@ func (c *Compiler) getMemoryLenValue(forceReload bool) ssa.Value {
 		memInstPtr := loadMemInstPtr.Return()
 
 		loadBufSizePtr := builder.AllocateInstruction()
-		loadBufSizePtr.AsLoad(memInstPtr, memoryInstanceBufSizeOffset, ssa.TypeI64)
+		if c.memoryShared {
+			sizeOffset := builder.AllocateInstruction().AsIconst64(memoryInstanceBufSizeOffset).Insert(builder).Return()
+			addr := builder.AllocateInstruction().AsIadd(memInstPtr, sizeOffset).Insert(builder).Return()
+			loadBufSizePtr.AsAtomicLoad(addr, 8, ssa.TypeI64)
+		} else {
+			loadBufSizePtr.AsLoad(memInstPtr, memoryInstanceBufSizeOffset, ssa.TypeI64)
+		}
 		builder.InsertInstruction(loadBufSizePtr)
 
 		ret = loadBufSizePtr.Return()
 	} else {
 		load := builder.AllocateInstruction()
-		load.AsExtLoad(ssa.OpcodeUload32, c.moduleCtxPtrValue, c.offset.LocalMemoryLen().U32(), true)
+		if c.memoryShared {
+			lenOffset := builder.AllocateInstruction().AsIconst64(c.offset.LocalMemoryLen().U64()).Insert(builder).Return()
+			addr := builder.AllocateInstruction().AsIadd(c.moduleCtxPtrValue, lenOffset).Insert(builder).Return()
+			load.AsAtomicLoad(addr, 8, ssa.TypeI64)
```

Comment:
I'm a bit confused on whether `8` is the correct value here. I see that the length seems to be `uint64`

https://github.com/tetratelabs/wazero/blob/main/internal/engine/wazevo/module_engine.go#L48

But below it is a `ssa.OpcodeUload32`

Reply:
good point, and I should fix the below so that I could use the normal 64bit load. 8 sounds correct for this